src/core/public/context/context.ts

```diff
@@ -0,0 +1,165 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { flatten } from 'lodash';
+import { PluginName } from '../../server';
+import { pick } from '../../utils';
+
+/**
+ * A function that returns a context value for a specific key of given context type.
+ *
+ * @remarks
+ * This function will be called each time a new context is built for a handler invocation.
+ *
+ * @param context - A partial context object containing only the keys for values provided by plugin dependencies
+ * @param rest - Additional parameters provided by the service owner of this context
+ * @returns The context value associated with this key. May also return a Promise.
+ *
+ * @public
+ */
+export type ContextProvider<
+  TContext extends {},
+  TContextName extends keyof TContext,
+  TProviderParameters extends any[] = []
+> = (
+  context: Partial<TContext>,
+  ...rest: TProviderParameters
+) => Promise<TContext[TContextName]> | TContext[TContextName];
+
+/**
+ * An object that handles registration of context providers and building of new context objects.
+ *
+ * @remarks
+ * Contexts providers are executed in the order they were registered. Each provider gets access to context values
+ * provided by any plugins that it depends on.
+ *
+ * @public
+ */
+export interface ContextContainer<TContext extends {}, TProviderParameters extends any[] = []> {
+  /**
+   * Register a new context provider. Throws an excpetion if more than one provider is registered for the same context
+   * key.
+   *
+   * @param contextName - The key of the {@link TContext} object this provider supplies the value for.
+   * @param provider - A {@link ContextProvider} to be called each time a new context is created.
+   * @param plugin - The plugin this provider is associated with. If `undefined`, provider gets access to all provided
+   *                 context keys.
+   * @returns The `ContextContainer` for method chaining.
+   */
+  register<TContextName extends keyof TContext>(
+    contextName: TContextName,
+    provider: ContextProvider<TContext, TContextName, TProviderParameters>,
+    plugin?: PluginName
+  ): this;
+
+  /**
+   * Create a new context.
+   *
+   * @param plugin - The plugin the context will be exposed to.
+   * @param baseContext - The initial context for the given handler.
+   * @param contextArgs - Additional parameters to call providers with.
+   * @returns A Promise for the new context object.
+   */
+  createContext(
+    plugin: PluginName,
+    baseContext: Partial<TContext>,
+    ...contextArgs: TProviderParameters
+  ): Promise<TContext>;
+}
+
+/** @internal */
+export class ContextContainerImpl<TContext extends {}, TProviderParameters extends any[] = []>
```

Comment:
In order to make these names a little more readable for non-English devs, I'd like to see us use full variables names. This may be a good example of switching the interface names around:

```ts
export class ContextContainer<TContext extends {}, TProviderParameters extends any[] = []>
  implements IContextContainer<TContext, TProviderParameters> {

// or:

export class ContextContainer<TContext extends {}, TProviderParameters extends any[] = []>
  implements ContextContainerBase<TContext, TProviderParameters> {
```

Reply:
Good thinking about the abbreviation being unclear. I really want the public name to be `ContextContainer` so I'm going to opt to rename `ContextContainerImpl` to `ContextContainerImplementation`. I'm not worried about this being too long since it's only referenced in 2 places.

Reply:
I'm cool with that. Another idea I had based on the class mocks issue was to make interfaces follow a `Contract` convention:

```js
implements ContextContainerContract<TContext, TProviderParameters>
```

But we can discuss this in the future.

Reply:
I've been using the `...Contract` convention. We use "contracts" like an `interface` but using a different word to `...Interface` means this could be a type or an interface so it doesn't matter if we're defining an interface or inferring a type with `PublicMethodsOf<...>`.